torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
+                from a call to :meth:`state_dict`.
         """
-
-        for key, value in state_dict["state"].items():
-            param = self._index_to_param[key]
-
-            # Populate the sharded optimizer state on the fly
+        for index, value in state_dict["state"].items():
+            param = self._index_to_param[index]
             if self._param_to_rank[param] != self.rank:
-                state_dict["state"][key] = None
+                # Clear any state irrelevant to this rank
+                state_dict["state"][index] = None
             else:
+                # Load the parameter state to the local optimizer
                 self.optim.state[param] = _recursive_copy_to_device(value, non_blocking=True, device=param.device)
 
         super().load_state_dict(state_dict)
 
-        # Sync with the optimizer param groups
+        # Sync the input state with the exposed and local optimizer states
         self._sync_param_groups(state_dict["param_groups"], self.param_groups)
         self._sync_param_groups(self.param_groups, self.optim.param_groups)
 
     def local_state_dict(self) -> Dict:
         r"""
-        Gets this rank's ``state_dict``.
+        Gets this rank's local ``state_dict``.
 
         Returns:
-            The state of the optimizer as a :class:`dict`.
-            It contains two entries:
-
-            * state - a dict holding current optimization state. Its content
-                differs between optimizer classes.
-            * param_groups - a dict containing all parameter groups
+            The state of the local optimizer as a :class:`dict`, containing
+            two entries:
+
+            * state
+                a dict holding the current optimization state, where the exact
+                contents depends on the optimizer class.
+            * param_groups
+                a list of parameter groups, each of which is a dict specifying
+                parameters to be optimized and group-specific optimization
+                options.
         """
         return self.optim.state_dict()
 
     def state_dict(self) -> Dict[str, Any]:
         r"""
-        Returns:
-            the last known global optimizer state, which consist of a list of
-            the shards.
+        Returns the last global optimizer state known to this rank.
 
         .. warning:
-            If the state has not been consolidated, this returns a shard's worth,
-            not the global state.
-
-        .. warning:
-            Returning the global state is limited to the replica which was
-            responsible for the consolidation. The state may also not be up to
-            date, depending on when :meth:`consolidate_state_dict` was last called.
+            If the state has not been consolidated to this rank, this raises a
+            runtime error, and even if it has, the state may not be up-to-date,
+            depending on when :meth:`consolidate_state_dict` was last called.
```

Comment:
I'm wondering if we need this interaction between state_dict and consolidate_state_dict? Every time we call `.step()` can we update a variable `needs_consolidation=True`. Then when state_dict is called, we first consolidate the state dicts if `needs_consolidateion=True`, cache the result and set `needs_consolidation=False`. This way the user only has to worry about calling `state_dict`.

Reply:
No, that would not work because of the collective needs. The issue here is that users might call `.state_dict()` on one rank only (in practice they often will), which means that you cannot consolidate at that point with any of the collective communication primitives. Something like RPC would alleviate that, but that's a bigger change, without this you need the collective consolidation before you can checkout (and doing this for every step is super slow, of course).